### PR TITLE
Create code coverage post-processing steps and integrate into PR Jobs

### PR DESCRIPTION
**Pull request by cfernald**
1. Switch to coverage files per-matrix to not publish coverage
2. Add steps to download and merge/filter all artifacts
3. Integrates steps into job templates.

Issue #84

### CODE REVIEW

**Comment by cfernald** [comment]
Opening as draft for reference, will likely be updated for tooling changes.

**Review by kuqin12 — APPROVED** [review]

**Review by apop5 — APPROVED** [review]

File: .sync/Version.njk [hunk 1]
```diff
@@ -36,4 +36,4 @@
 {% set latest_mu_release_branch = "release/202208" %}
 
 {# The version of the fedora-35-build container to use. #}
-{% set linux_build_container = "ghcr.io/tianocore/containers/fedora-37-build:14d2aba" %}
+{% set linux_build_container = "ghcr.io/tianocore/containers/fedora-37-build:f1c7a20" %}
```
Comment (makubacki) [inline .sync/Version.njk:39]: Please hold until Monday for me to look then.

File: Jobs/PrGate.yml [hunk 2]
```diff
@@ -46,7 +47,7 @@ parameters:
 - name: linux_container_image
   displayName: Linux Container Image
   type: string
-  default: 'ghcr.io/tianocore/containers/fedora-35-build:5b8a008'
+  default: 'ghcr.io/tianocore/containers/fedora-37-build:f1c7a20'
```
Comment (makubacki) [inline Jobs/PrGate.yml:50]: I ultimately want to get to the point where we track the exact container image (by SHA) in one place. Right now that is Version.njk.

My thought was:

1. Any pipeline that cares to use a specific version should pass it in as a parameter (Version.njk is one way to get the value and do that)
2. The default requires less maintenance over time

With that in mind, can this just be `"fedora-37-build:latest"` for now? As long as latest works as expected in the tianocore container repo (it does not always depending on how you manage images).

File: Jobs/PrGate.yml [hunk 3]
```diff
@@ -110,3 +111,15 @@ jobs:
...
         install_tools: ${{ not(parameters.container_build) }}
+        extra_build_args: CODE_COVERAGE=TRUE
+
+  - job: PublishCoverage
+    dependsOn: Build
+    workspace:
+      clean: all
+
+    pool:
+      vmImage: 'windows-2022'
```
Comment (makubacki) [inline Jobs/PrGate.yml:122]: If this doesn't have VS toolchain dependencies (it does not) then I suggest to use "`windows-latest`". It will be one less place to fall behind and cause problems in the future when "`windows-2022`" has brown outs and is eventually deprecated.

File: Steps/PublishCodeCoverage.yml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+parameters:
+- name: include_paths
+  displayName: Tool Chain (e.g. VS2022)
+  type: string
+  default: ''
+- name: exclude_paths
+  displayName: Tool Chain (e.g. VS2022)
+  type: string
+  default: ''
```
Comment (makubacki) [inline Steps/PublishCodeCoverage.yml:16]: `include_paths` and `exclude_paths` are not used in this file. Also the `displayName` comment seems unrelated.

Comment (cfernald) [reply Steps/PublishCodeCoverage.yml:16]: Yes, I decided to drop this in one of the stages and forgot to remove the parameters

File: Jobs/PrGate.yml (hunk 2 above)
Comment (cfernald) [reply Jobs/PrGate.yml:50]: My biggest concern with the latest is that if we support forked branches, they would keep getting the new container but with all the old build definitions. It's possible there could be incompatibilities there we would have to concern ourselves with like dropping tools that are no longer used, etc.

Comment (makubacki) [reply Jobs/PrGate.yml:50]: Does any code depend on the default value right now? Ultimately, I think all template users should pass this value since they need to account for changes in the image. For example, this could be an empty default and we raise a pipeline error if it remains empty.

If the repo opts in to receive its "leaf" pipelines from file sync, they get it automatically (from Version.njk). Otherwise, they can pull Version.njk down in their pipeline run and extract the value.

Comment (makubacki) [reply Jobs/PrGate.yml:50]: > Is pulling version.njk feasible for Azure dev-ops repos?

Sure. Here's a quick curl example:
```bash
curl -s https://raw.githubusercontent.com/microsoft/mu_devops/main/.sync/Version.njk | grep linux_build_container | cut -d '"' -f 2
```

What I'm proposing is that we do this one way. Either (1) require template users to specify what container image they want or (2) always specify the container image in the pipeline.

By this being a template parameter, that means (1) has been chosen. So, the complexity associated with testing that is inherited. (2) brings more complexity that I don't think is needed if (1) is supported. I think (1) makes sense because it allows individual repos to control their build environment.

Then, it becomes a testing problem. Which is an issue (#23). But, because this repo produces versioned release, dependent repos do not need to update what mu_devops they use until they can verify their chosen container image works with the mu_devops version they are updating to and that can be confirmed in their CI build process.

Comment (cfernald) [reply Jobs/PrGate.yml:50]: Yeah, I suppose I see this as a convenience. For all the synced repos it's moot really since they get their top-level pipeline files from this repo anyways so the burden of testing really doesn't change one way or the other in my opinion. For all the non-synced repos this allows an easy way to just stay in lock-step with dev_ops. All such repos will need to manually roll their dev_ops version anyways, so whether they need to validate the container separately or not seems to make no difference testing-wise in my mind. This just provides an easy solution if they don't need container customizations.

Either way, I don't think there are any non-synced repros that use containers anyways, so i'm happy to just push this to the azure_pipelines/custom/* files using version.njk if that work for you.

File: Steps/InstallCoverageTools.yml [hunk 5]
```diff
@@ -0,0 +1,14 @@
+## @file
```
[the remark this answers is not in the document]
Comment (cfernald) [reply Steps/InstallCoverageTools.yml:1]: Agreed, this type of documentation I think will live in the basecore/mu documentation though.

**Review by makubacki — APPROVED** [review]
